sound/soc/sof/ipc.c

```diff
@@ -672,9 +672,31 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
 	struct sof_ipc_fw_version *v = &ready->version;
 	struct sof_ipc_ctrl_data_params sparams;
+	struct snd_sof_widget *swidget;
+	bool widget_found = false;
 	size_t send_bytes;
 	int err;
 
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (swidget->comp_id == scontrol->comp_id) {
+			widget_found = true;
+			break;
+		}
+	}
+
+	if (!widget_found) {
+		dev_err(sdev->dev, "error: can't find widget with id %d\n", scontrol->comp_id);
+		return -EINVAL;
+	}
+
+	/*
+	 * Volatile controls should always be part of static pipelines and the widget use_count
+	 * would always be > 0 in this case. For the others, just return the cached value if the
+	 * widget is not set up.
+	 */
+	if (!swidget->use_count)
+		return 0;
+
```

Comment:
Out of curiosity, in which cases would we have widgets that belong to multiple pipelines?

I can only think of bi-directional DAIs, are there other examples?

Reply:
even mixers/mux. basically the list of connected DAPm widgets for 2 PCMs that share a mixer will both contain the mixer. So if one PCM is started, the second one should skip setting it up and oncversely while freeing, if one is using it, we should skip freeing it

Reply:
Mux and mixers will have their sources in a set of pipelines but their sink is in a different one. Does this make the mux and mixers "belong" to different pipelines?  What does "belong" mean really?

Reply:
belong is probably misleading yout into thinking that the widget is part of both pipelines in topology. WHat I really mean is a widget like the mixer will "belong" to the list of of connected DAPM widgets for both the PCMs that use the mixer widget.